### DISCUSSION TITLE
iPXE hang while 'disconnecting existing drivers'

### DISCUSSION THREAD

**Discussion by codwazny** [ipxe/ipxe#287] [Q&A] [answered]
I have an issue on some of our servers where we have iPXE hangs during initialising devices phase. After a few minutes of hanging, they will reboot. The issue does not happen 100% of the time which makes it even harder to figure out. Occasionally it will go through without any issues but often times it will get stuck and hang.

To try and give the best logging back I recompiled iPXE with: 
```
debug="device,runtime,pci,init,nii,snp,snpnet,snponly,efi_watchdog,efi_image,image_cmd,image,efi_autoboot,efi_timer,efi_driver:3,efi_debug,efi_init,efi_pci,efi_wrap,intel,intelxl"
```

Right before the hang the last line of logging I get to see is:
```
EFIDRV PciRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x0)/MAC(F8F21E7A0800,0x1) disconnecting existing drivers
```

My understanding is that the hang is coming from a call to UEFI which should be clearing all the drivers out. That code I believe is right here: https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_driver.c#L455-L456

Is there something I should be digging in further here? Is this an issue with iPXE or something else? Could the issue lie in the BIOS or NIC firmware? It seems like it's hanging in a UEFI call?

Extra details:
Full iPXE log of hang: https://gist.github.com/codwazny/386278ae65b4a73bb9f449ef3e888b27
Full iPXE log of working boot: https://gist.github.com/codwazny/81e8848f03652b6bee99284a6ebe22e2

The servers:
Dell R340 w/ X710 card (I often see this same issue on other Dell models as well). The server is running latest BIOS (2.4.1) and latest NIC firmware (19.5.12).

Have tried both snponly and ipxe builds but both hang as they both seem to get stuck on disconnecting this same driver.

**Comment by mcb30** [comment, marked as answer]
Looking at the diff of your "hang" vs. "working" boots up to the point of `disconnecting existing drivers`, almost all of the output is identical (apart from pointer values changing slightly).  There is only one non-trivial difference:

```diff
 EFI image base address 0x59133000
 EFI image device path PciRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x0)/MAC(F8F21E7A0800,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000)
-EFI PciRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x0)/MAC(F8F21E7A0800,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000) found autoboot link-layer address:
+EFI found autoboot link-layer address:
 00000000 : f8 f2 1e 7a 08 00                               : ...z..
```

That change in the debug message (which happened to take place in commit https://github.com/ipxe/ipxe/commit/a3f1e8fb6) indicates that your "hang" and "working" outputs **do not come from the same source code**.  Your "working" output is from a newer version of iPXE than your "hang" output.

Could you please retest with the current codebase and raise an issue (via https://github.com/ipxe/ipxe/issues) complete with updated debug logs if the problem still occurs even with the latest code?

Thanks.

**Reply by codwazny** [reply]
Thanks for the reply. I'll get a repro from the same exact build. I thought these came from the same rev but they might have been built slightly different with more debug enabled. I'll be sure to get one with the exact same debug flags and move it over to issues.

**Reply by mcb30** [reply]
They are built with the same debug options but are definitely not from the same git commit (see explanation above).

**Reply by codwazny** [reply]
Thanks. I was able to figure out the different rev issue. I opened https://github.com/ipxe/ipxe/issues/291 with updated logs from the exact same binary built on 1192edf394aaebde666c6aafda1c8b782b08a845. We can continue the discussion there!
